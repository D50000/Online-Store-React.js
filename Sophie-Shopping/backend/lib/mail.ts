import { createTransport, getTestMessageUrl } from "nodemailer";

const transporter = createTransport({
  host: "smtp.ethereal.email",
  port: 587,
  auth: {
    user: "donny.beatty34@ethereal.email",
    pass: "shxZASXrt7NC2paQTv",
  },
});

function createSimpleEmail(text: string): string {
  return `
    <div className="email" style="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
    ">
      <h2>Hello There!</h2>
      <p>${text}</p>
      <p>Hi, Sophie Shopping's User !!! </p>
    </div>
  `;
}

export interface MailResponse {
  accepted?: string[] | null;
  rejected?: null[] | null;
  envelopeTime: number;
  messageTime: number;
  messageSize: number;
  response: string;
  envelope: Envelope;
  messageId: string;
}

export interface Envelope {
  from: string;
  to?: string[] | null;
}

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  const info = (await transporter.sendMail({
    to,
    from: "test@test.com",
    subject: "Your password reset token!",
    html: createSimpleEmail(`Your Password Reset Token Is Here!
    <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click Here to reset</a>
  `),
  })) as MailResponse;

  if (process.env.MAIL_USER.includes("ethereal.email")) {
    console.log(`ðŸ’Œ Message Sent!  Preview it at ${getTestMessageUrl(info)}`);
  }
}
