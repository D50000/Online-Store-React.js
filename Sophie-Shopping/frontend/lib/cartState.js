import { createContext, useContext, useState } from 'react';
import PropTypes from 'prop-types';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // Custom provider can access it via the consumer.

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// Make the custom hook for these functions.
function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };

CartStateProvider.propTypes = {
  children: PropTypes.any,
};
