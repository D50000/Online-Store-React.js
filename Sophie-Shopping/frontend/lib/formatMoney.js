export default function formatMoney(amount = 0) {
  const options = {
    style: 'currency',
    currency: 'TWD',
    minimumFractionDigits: 0,
  };

  const formatter = Intl.NumberFormat('zh-US', options);

  return formatter.format(amount);
}
