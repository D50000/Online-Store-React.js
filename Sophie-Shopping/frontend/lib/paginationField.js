import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // Setup Apollo keys myself.
    read(existing = [], { args, cache }) {
      // console.log({ existing, args, cache });
      const { skip, first } = args;

      // Read the detail of items on the page from the cache.
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // Check if we have existing items.
      const items = existing.slice(skip, skip + first).filter((item) => item);

      // On the last page.
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      // No item, fetch it from the DB.
      if (items.length !== first) {
        return false;
      }
      if (items.length) {
        return items;
      }

      // Fallback to network
      return false;
    },
    // Apollo fetch back the data will run merge()
    merge(existing, incoming, { args }) {
      const { skip, first } = args;

      // Calculate the item need to update.
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; i + 1) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}
