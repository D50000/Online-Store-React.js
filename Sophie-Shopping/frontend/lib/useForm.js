import { useState } from 'react';

export default function useForm(initial = {}) {
  const [inputs, setInputs] = useState(initial);

  function handleChange(e) {
    // Handle the data type that pass in.
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      value[0] = e.target.files;
    }
    setInputs({
      ...inputs,
      // Dynamic update the key and value
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const emptyState = Object.fromEntries(
      Object.entries(inputs).map(([key]) => [key, ''])
    );
    setInputs(emptyState);
  }

  // Return the initialState and SetStateAction()
  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
