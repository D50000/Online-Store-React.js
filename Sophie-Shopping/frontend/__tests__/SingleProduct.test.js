import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';

import SingleProductPage, {
  SINGLE_ITEM_QUERY,
} from '../components/SingleProduct';
import { fakeItem } from '../lib/testUtils';

const product = fakeItem();

// Arrange, mock the query data
const mocks = [
  {
    request: {
      query: SINGLE_ITEM_QUERY,
      variables: {
        id: '123',
      },
    },
    result: {
      data: {
        Product: product,
      },
    },
  },
];

describe('<Single Product>', () => {
  it('render with proper data', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProductPage id="123" />
      </MockedProvider>
    );
    // Act
    // Use async/await function to handle the query response.
    await screen.findByTestId('singleProduct');
    debug();
    // Assert
    expect(container).toMatchSnapshot();
  });

  it('Errors message when no item', async () => {
    // Arrange, mock the query data
    const errorMocks = [
      {
        request: {
          query: SINGLE_ITEM_QUERY,
          variables: {
            id: '123',
          },
        },
        result: {
          errors: [{ message: 'No item found ....' }],
        },
      },
    ];
    const { container, debug } = render(
      <MockedProvider mocks={errorMocks}>
        <SingleProductPage id="123" />
      </MockedProvider>
    );
    // Act
    // Use async/await function to handle the query response.
    await screen.findByTestId('graphql-error');
    debug();
    // Assert
    expect(container).toHaveTextContent('No item found ....');
  });
});
