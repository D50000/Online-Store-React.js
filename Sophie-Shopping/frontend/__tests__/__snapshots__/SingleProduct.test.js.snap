// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Single Product> render with proper data 1`] = `
<div>
  <div
    class="SingleProduct__ProductStyles-sc-4lzk54-0 PvcsM"
    data-testid="singleProduct"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
